core/include/msg.h

```diff
@@ -11,14 +11,141 @@
- * msg_init_queue(msg_t* array, int num). Messages sent to a thread with a non
- * full message queue are never dropped * and the sending never blocks. Threads
- * with a full message queue behaves like in synchronous mode.
+ * Messages
+ * ========
+ * IPC messages consist of a sender PID, a type, and some content. The sender
+ * PID will be set by the IPC internally and is not required to be set by the
+ * user. The type helps the receiver to multiplex different message types and
+ * should be set to a system-wide unique value. The content can either be
+ * provided as a 32-bit integer or a pointer.
+ *
+ * Blocking vs non-blocking
+ * ========================
+ * Messages can be sent and received blocking and non-blocking. Both can be
+ * used combined: A message send while blocking the sender thread can be
+ * received with the non-blocking variant and vice-versa.
+ *
+ * Blocking IPC
+ * ------------
+ * For the blocking variant use @ref msg_send() or @ref msg_receive()
+ * respectively.
+ *
+ * Additionally, one can use @ref msg_send_receive() to simultaneously block
+ * the sending thread and expect a response response from the receiving thread.
+ * In this case, the receiving thread must use @ref msg_reply() to reply to the
+ * message of the sender thread.
+ *
+ * ~~~~~~~~~~~~~~~~~~~~~~~~ {.c}
+ * #include <inttypes.h>
+ * #include <stdio.h>
+ *
+ * #include "msg.h"
+ * #include "thread.h"
+ *
+ * static kernel_pid_t rcv_pid;
+ * static char rcv_stack[THREAD_STACKSIZE_DEFAULT];
+ *
+ * static void *rcv(void *arg)
+ * {
+ *     msg_t msg_req, msg_resp;
+ *
+ *     (void)arg;
+ *     while (1) {
+ *         msg_receive(&msg_req);
+ *         msg_resp.content.value = msg_req.content.value + 1;
+ *         msg_reply(&msg_req, &msg_resp);
+ *     }
+ *     return NULL;
+ * }
+ *
+ * int main(void)
+ * {
+ *     msg_t msg_req, msg_resp;
+ *
+ *     msg_resp.content.value = 0;
+ *     rcv_pid = thread_create(rcv_stack, sizeof(rcv_stack),
+ *                             THREAD_PRIORITY_MAIN - 1, 0, rcv, NULL, "rcv");
+ *     while (1) {
+ *         msg_req.content.value = msg_resp.content.value;
+ *         msg_send_receive(&msg_req, &msg_resp, rcv_pid);
+ *         printf("Result: %" PRIu32 "\n", msg_resp.content.value);
+ *     }
+ *     return 0;
+ * }
+ * ~~~~~~~~~~~~~~~~~~~~~~~~
+ *
+ * Non-blocking IPC
+ * ----------------
+ * For the non-blocking variant use @ref msg_try_send() or
+ * @ref msg_try_receive() respectively. If a message is send in synchronous
+ * mode or the message queue (see below) of the receiving thread is full
+ * messages send this way will be dropped.
+ *
+ * As an example you can use the example on asynchronous IPC below without the
+ * queue.
+ *
+ * Synchronous vs Asynchronous
+ * ===========================
+ * RIOT's IPC supports both synchronous and asynchronous IPC.
+ *
+ * Synchronous IPC is the default mode and one does not need to do anything to
```

Comment:
I decided against subsections, because synchronous IPC was just one sentence, but I can use it anyway.